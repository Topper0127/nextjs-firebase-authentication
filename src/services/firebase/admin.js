import * as firebaseAdmin from 'firebase-admin';

import firebaseServiceAccountKey from '../../../firebaseServiceAccountKey.json';

if (!firebaseAdmin.apps.length) {
  firebaseAdmin.initializeApp({
    credential: firebaseAdmin.credential.cert(
      firebaseServiceAccountKey
    ),
    databaseURL: process.env.FIREBASE_DATABASE_URL,
  });
}

export default firebaseAdmin;
