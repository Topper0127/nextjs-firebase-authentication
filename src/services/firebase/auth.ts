export const doPasswordReset = (email: string) => {
  // auth.sendPasswordResetEmail(email);
};

export const doPasswordUpdate = (password: string) => {
  // if (auth.currentUser) {
  //   return auth.currentUser.updatePassword(password);
  // } else {
  //   return Promise.reject(new Error('No authenticated user.'));
  // }
};
