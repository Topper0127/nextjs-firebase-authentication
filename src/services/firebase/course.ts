import * as firebaseAdminVanilla from 'firebase-admin';

import firebaseAdmin from '@services/firebase/admin';

import { COURSE } from '@data/course-keys-types';
import { BUNDLE } from '@data/bundle-keys-types';

export const createCourse = async ({
  uid,
  courseId,
  bundleId,
  amount,
  paymentType,
  coupon,
}: {
  uid?: string;
  courseId: COURSE;
  bundleId: BUNDLE;
  amount: number;
  paymentType: string;
  coupon: string;
}) =>
  await firebaseAdmin
    .database()
    .ref(`users/${uid}/courses`)
    .push()
    .set({
      courseId: courseId,
      packageId: bundleId,
      invoice: {
        createdAt:
          firebaseAdminVanilla.database.ServerValue.TIMESTAMP,
        amount,
        licensesCount: 1,
        currency: 'USD',
        paymentType,
        coupon,
      },
    });
