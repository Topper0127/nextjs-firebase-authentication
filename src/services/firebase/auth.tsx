import { auth } from './firebase';

export const doCreateUserWithEmailAndPassword = (
  email: string,
  password: string
) => auth.createUserWithEmailAndPassword(email, password);

export const doSignInWithEmailAndPassword = (
  email: string,
  password: string
) => auth.signInWithEmailAndPassword(email, password);

export const doSignOut = () => auth.signOut();

export const doPasswordReset = (email: string) =>
  auth.sendPasswordResetEmail(email);

export const doPasswordUpdate = (password: string) => {
  if (auth.currentUser) {
    auth.currentUser.updatePassword(password);
  }
};
