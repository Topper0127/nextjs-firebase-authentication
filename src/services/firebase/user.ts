import { db } from './firebase';

export const doCreateUser = (
  id: string,
  username: string,
  email: string
) =>
  db.ref(`users/${id}`).set({
    username,
    email,
  });
