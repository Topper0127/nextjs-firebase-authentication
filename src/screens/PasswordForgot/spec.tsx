import renderer from 'react-test-renderer';
import waitForExpect from 'wait-for-expect';

import { message } from 'antd';

import * as authService from '@services/firebase/auth';

import { PasswordForgotFormEnhanced } from '.';

describe('PasswordForgotFormEnhanced', () => {
  let component: any;
  let getComponent: any;
  let defaultProps: any;

  beforeEach(() => {
    defaultProps = {};

    getComponent = (props: any = defaultProps) => (
      <PasswordForgotFormEnhanced {...props} />
    );

    component = renderer.create(getComponent());
  });

  beforeEach(() => {
    message.loading = jest.fn();
    message.error = jest.fn();
    message.success = jest.fn();
  });

  it('renders', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('resets a password', () => {
    const email = 'example@example.com';

    beforeEach(() => {
      component.root
        .findByProps({ id: '/password-forgot_email' })
        .props.onChange({ target: { value: email } });
    });

    it('with success', async () => {
      const spy = jest
        .spyOn(authService, 'doPasswordReset')
        .mockImplementation(() => Promise.resolve());

      component.root
        .findByType('form')
        .props.onSubmit({ preventDefault: () => {} });

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(email);

      expect(message.loading).toHaveBeenCalledTimes(1);

      await waitForExpect(() => {
        expect(message.error).toHaveBeenCalledTimes(0);
        expect(message.success).toHaveBeenCalledTimes(1);
      });
    });

    it('with error', async () => {
      const spy = jest
        .spyOn(authService, 'doPasswordReset')
        .mockImplementation(() =>
          Promise.reject(new Error('Wrong password.'))
        );

      component.root
        .findByType('form')
        .props.onSubmit({ preventDefault: () => {} });

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(email);

      expect(message.loading).toHaveBeenCalledTimes(1);

      await waitForExpect(() => {
        expect(message.error).toHaveBeenCalledTimes(1);
        expect(message.success).toHaveBeenCalledTimes(0);
      });
    });
  });
});
