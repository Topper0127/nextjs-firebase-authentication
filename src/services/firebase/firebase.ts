import app from 'firebase/app';

import 'firebase/database';
import 'firebase/auth';
import 'firebase/storage';

const config = {
  apiKey: process.env.FIREBASE_API_KEY,
  authDomain: process.env.FIREBASE_AUTH_DOMAIN,
  databaseURL: process.env.FIREBASE_DATABASE_URL,
  projectId: process.env.FIREBASE_PROJECT_ID,
  storageBucket: process.env.FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.FIREBASE_APP_ID,
};

if (!app.apps.length) {
  app.initializeApp(config);
}

const db = app.database();
const auth = app.auth();
const storage = app.storage();

// TODO
// As httpOnly cookies are to be used, do not persist any state client side.
// firebase.auth().setPersistence(firebase.auth.Auth.Persistence.NONE);

export default app;

export { db, auth, storage };
