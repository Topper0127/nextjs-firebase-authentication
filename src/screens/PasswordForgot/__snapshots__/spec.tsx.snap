// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordForgotFormEnhanced renders 1`] = `
<form
  className="ant-form ant-form-horizontal"
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item Item__StyledFormItem-sc-13gzk8a-0 lhvvBo"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-6"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="/password-forgot_email"
        onClick={[Function]}
        title="E-mail"
      >
        E-mail
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper ant-col-xs-24 ant-col-sm-18"
      style={Object {}}
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "name": "email",
              }
            }
            data-__meta={
              Object {
                "name": "email",
                "originalProps": Object {
                  "type": "text",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "The input is not valid E-mail!",
                    "type": "email",
                  },
                  Object {
                    "message": "Please input your E-mail!",
                    "required": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "The input is not valid E-mail!",
                        "type": "email",
                      },
                      Object {
                        "message": "Please input your E-mail!",
                        "required": true,
                      },
                    ],
                    "trigger": Array [
                      "onBlur",
                    ],
                  },
                ],
                "validateFirst": true,
                "validateTrigger": "onBlur",
                "valuePropName": "value",
              }
            }
            id="/password-forgot_email"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value=""
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item Item__StyledFormItem-sc-13gzk8a-0 lhvvBo"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-control-wrapper ant-col-sm-24"
      style={Object {}}
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <button
            className="ant-btn StretchedButton-sc-1skj5p1-0 gqwrSP ant-btn-primary"
            onClick={[Function]}
            type="submit"
          >
            <span>
              Reset Password
            </span>
          </button>
        </span>
      </div>
    </div>
  </div>
</form>
`;
